support more GeoJSON geometries

nevi-me opened the pull request:
ref #3, it's incomplete as I haven't added support for `GeometryCollection`. The idea with supporting `GeometryCollection` is that we'll unbundle `modules.export()` further into separate functions for lines and polygons, then loop through the features in a geometry collection of feature collection. So I need some input before going ahead with that.

I'll catch up to @frankrowe 's work before we merge this PR.
- [x] `MultiLineString`
- [x] `MultiPolygon`
- [x] `GeometryCollection`
- [x] passthrough for other types
- [x] error checking (@frankrowe)
- [x] update docs to include supported inputs

nevi-me committed:
support more GeoJSON geometries

nevi-me commented:
just noticed that this is related to and clashes with #2

tchannel commented:
@nevi-me looks like you've got a lot more covered here than I did in #2. I'll close that and review what you've got here

nevi-me commented:
Thanks @tchannel, I've fallen a bit behind during the week, but will complete this on the weekend

nevi-me committed:
start unbundling module.exports to support more GeoJSON types

nevi-me committed:
update tests

nevi-me committed:
Merge branch 'master' of https://github.com/nevi-me/turf-simplify into any-input

Conflicts:
	index.js
	test/test.js

nevi-me committed:
FeatureCollection now works, and fixed issues post-merging with 'over-simplification'

nevi-me committed:
support geometrycollection, and some minor fixes

nevi-me committed:
revert tolerance to what it was on tests

nevi-me committed:
update benchmarks

```
turf-simplify#LineString x 183,467 ops/sec ±0.34% (101 runs sampled)
turf-simplify#MultiLineString x 55,560 ops/sec ±0.45% (99 runs sampled)
turf-simplify#Polygon x 36,431 ops/sec ±0.61% (100 runs sampled)
turf-simplify#MultiPolygon x 1,552 ops/sec ±0.56% (99 runs sampled)
turf-simplify#SimplePolygon x 10,075 ops/sec ±0.57% (98 runs sampled)
turf-simplify#FeatureCollection x 12,499 ops/sec ±0.22% (98 runs sampled)
turf-simplify#GeometryCollection x 12,109 ops/sec ±0.36% (99 runs sampled)
turf-simplify#Argentina x 27,408 ops/sec ±0.32% (101 runs sampled)

```

nevi-me commented:
I'm done @tchannel @morganherlocker. If it walks and talks GeoJSON, it'll be simplified. I merged in the validations that @frankrowe made. Commit's ready for review. There are some parts that I'd like feedback on, but too tired to list them now, we'll see them as we go.

Previous benchmarks:

```
turf-simplify#LineString x 180,827 ops/sec ±0.29% (100 runs sampled)
turf-simplify#Polygon x 37,203 ops/sec ±0.55% (98 runs sampled)
turf-simplify#Argentina x 27,475 ops/sec ±0.21% (99 runs sampled)
```

Current benchmarks

```
turf-simplify#LineString x 183,467 ops/sec ±0.34% (101 runs sampled)
turf-simplify#MultiLineString x 55,560 ops/sec ±0.45% (99 runs sampled)
turf-simplify#Polygon x 36,431 ops/sec ±0.61% (100 runs sampled)
turf-simplify#MultiPolygon x 1,552 ops/sec ±0.56% (99 runs sampled)
turf-simplify#SimplePolygon x 10,075 ops/sec ±0.57% (98 runs sampled)
turf-simplify#FeatureCollection x 12,499 ops/sec ±0.22% (98 runs sampled)
turf-simplify#GeometryCollection x 12,109 ops/sec ±0.36% (99 runs sampled)
turf-simplify#Argentina x 27,408 ops/sec ±0.32% (101 runs sampled)
```

nevi-me commented:
Anyone have some capacity to look at this in the near future? @tcql @morganherlocker

tcql commented:
@nevi-me I'll try and get a look at it this weekend

nevi-me commented:
ta, would really appreciate it.

On 11 September 2015 at 20:38, Tim Channell <notifications@github.com>
wrote:

> @nevi-me <https://github.com/nevi-me> I'll try and get a look at it this
> weekend
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Turfjs/turf-simplify/pull/7#issuecomment-139622854>.
>

tcql commented on index.js:
```diff
@@ -51,44 +51,104 @@ var simplify = require('simplify-js');
+/*
+* returns true if ring is a triangle
+*/
+function checkTriangle(ring) {
```
I'd vote for renaming to `checkValidity` or something. Reading `checkTriangle` down below was misleading; I spent a few minutes trying to figure out why you would want to force a simplified polygon to be a triangle

tcql commented on index.js:
```diff
@@ -97,3 +157,50 @@ function simpleFeature (geom, properties) {
+        return [coords.x, coords.y];
+      });
+    }
+    if (simpleRing.length === 3) {
```
Because of the triangle checks above, L192-197 could be removed: 

```
// Not needed, because on line 198, we're ensuring the first and last element are equal
if (simpleRing.length === 3) {
      simpleRing.push(simpleRing[0]);
}

// Not needed because you're already handling shapes that are smaller than a triangle above
for (var i = 0; i < 4; i++) {
  if (!simpleRing[i]) simpleRing.push(simpleRing[0]);
}
```

tcql commented:
also, the first and last points of `test/fixtures/in/polygon.geojson` don't match

nevi-me committed:
rename testTriangle >> testValidity

`testValidity` is a clearer description to use. Renamed all instances in the file of `testTriangle`. Also removed dead code that's redundant due to the validity check

nevi-me committed:
correct invalid GeoJSON Polygon test input

The Polygon was invalid because the first set of coordinates was not the same as the last.

nevi-me commented:
@tcql fixed, thanks for picking up the redundant code. I renamed `testTriangle` to `testValidity` as suggested, removed the redundant code, and fixed the polygon.

frankrowe commented:
:+1:

tcql commented:
great work @nevi-me, thanks!

nevi-me commented:
Thanks, I'll def be contributing more in future!
